### DISCUSSION TITLE
Duplication of upstream Dialer code

### DISCUSSION THREAD

**Discussion by rossigee** [dmacvicar/terraform-provider-libvirt#1134] [Q&A] [answered]
Hi @dmacvicar,

I'm using a `qemu+tls` remote host, but the plugin just hangs indefinitely after the 'Connect'. While debugging I notice you have re-implemented all the Dialer creation code that seems to already exist in the upstream library...

https://github.com/digitalocean/go-libvirt/tree/main/socket/dialers

Is there a specific reason for this?

Could (/should?) this plugin be optimised to use the upstream functions instead?

Cheers.

**Comment by dmacvicar** [comment, marked as answer]
The reason is that the connection URIs where implemented first in our terraform provider.

https://github.com/digitalocean/go-libvirt/issues/139

The upstream implementation is not very old, and it is very inspired in ours. Thanks to @zaneb :tada: 

The obvious path is to switch to the upstream implementation, but in doing so, we need to see if features would be missing. I am mostly worried about all the ssh features (eg support for local config) and bastion support.

We may find a way to just wrap the upstream one and use our implementation for some of the transports.

**Reply by rossigee** [reply]
Thanks for the details :+1:

**Reply by dmacvicar** [reply]
Related: https://github.com/dmacvicar/terraform-provider-libvirt/discussions/1164

**Comment by zaneb** [comment]
What Duncan said is correct. FWIW the bug is almost certainly the one described [here](https://github.com/digitalocean/go-libvirt/issues/89#issuecomment-1607300636) - the libvirt tls protocol requires you to read a byte from the socket before you do anything else. It appears from that bug report that older versions of go-libvirt somehow masked that issue until mid-2019.
